harness/tests/cli/test_auth.py

```diff
@@ -43,9 +43,255 @@ def test_auth_with_store(requests_mock: requests_mock.Mocker, user: Optional[str
+        ScenarioSet("n", "*", "ye", "*", "y", "y", Check("cache")),
+        # Explicit user and password, DET_USER set but DET_USER_TOKEN not set.  DET_PASS is ignored.
+        ScenarioSet("y", "y", "y", "*", "n", "y", Check("cache")),
+        ScenarioSet("y", "y", "y", "*", "n", "n", Login("user", "req_pass")),
+        ScenarioSet("y", "y", "y", "*", "n", "x", Check("expired"), Login("user", "req_pass")),
+        # Explicit user but no password, DET_USER and DET_PASS set, and DET_USER_TOKEN unset.
+        # DET_PASS still ignored since DET_USER/DET_PASS are meant to be processed as a unit, and
+        # an explicitly-requested username overrides that unit.
+        ScenarioSet("y", "n", "y", "y", "n", "y", Check("cache")),
+        ScenarioSet("y", "n", "y", "y", "n", "n", Login("user", "prompt_pass")),
+        ScenarioSet("y", "n", "y", "y", "n", "x", Check("expired"), Login("user", "prompt_pass")),
+        # Explicit user but no password, DET_USER set but no other env.  DET_USER is ignored.
+        ScenarioSet("y", "n", "y", "n", "n", "y", Check("cache")),
+        ScenarioSet("y", "n", "y", "n", "n", "n", Login("user", "prompt_pass")),
+        ScenarioSet("y", "n", "y", "n", "n", "x", Check("expired"), Login("user", "prompt_pass")),
+        # Nothing explicit; DET_USER and DET_PASS are set, DET_USER_TOKEN is unet.
+        # Cache continues to work where it matches DET_USER, and there are no password prompts.
+        ScenarioSet("n", "*", "y", "y", "n", "y", Check("cache")),
+        ScenarioSet("n", "*", "y", "y", "n", "n", Login("user", "env_pass")),
+        ScenarioSet("n", "*", "y", "y", "n", "x", Check("expired"), Login("user", "env_pass")),
+        # Nothing explicit; and DET_USER is ignored without either DET_PASS or DET_USER_TOKEN.
+        ScenarioSet("n", "n", "*", "n", "n", "n", Login("determined", "")),
+        # the username is taken from the cache but password must be provided again
+        ScenarioSet("n", "n", "*", "n", "n", "x", Check("expired"), Login("user", "prompt_pass")),
+        ScenarioSet("n", "n", "*", "n", "n", "y", Check("cache")),
+        # If password is explicit but username is not, we fall back to the default username.
+        ScenarioSet("n", "y", "n", "n", "n", "n", Login("determined", "req_pass")),
+        # Other password-but-not-username cases are governed by other effects (also see above).
+        ScenarioSet("n", "y", "*", "*", "*", "y", Check("cache")),
+    ],
+)
+@mock.patch("determined.common.api.authentication._is_token_valid")
+@mock.patch("determined.common.api.authentication.do_login")
+@mock.patch("getpass.getpass")
+@mock.patch("determined.common.api.authentication.TokenStore.get_active_user")
+@mock.patch("determined.common.api.authentication.TokenStore.get_token")
+@mock.patch("determined.common.api.authentication.TokenStore.drop_user")
+@mock.patch("determined.common.api.authentication.TokenStore.set_token")
+def test_login_scenarios(
```

Comment:
What fixtures can make this function easier? It's a lot of set-up and tear-down.

Reply:
I have made some good improvements here